manage.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 8/14/25               \" DATE
.Dt manage.sh 1
.Os Linux
.Sh NAME
.Nm manage.sh
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line.
.\" Use .Nm macro to designate other names for the documented program.
.Nd Manage files in deeply nested directories
.Sh SYNOPSIS
.Nm
[OPTION]... SOURCE
.Sh DESCRIPTION
Perform operations on SOURCE directory.
.Pp
Multiple operations can be performed in a single invocation. Flags will be
evaluated in the order they are received.
.Pp
The program will automatically assume that the final parameter is not related to
any other flag, and moreover represents the source of the files.
.Bl -tag -width Ds
.It Fl f, --flatten-directory
Copies all nested files to a flat, specified output directory, set with
.Fl o
.It Fl e, --list-extensions
Traverses all files in SOURCE and prints filename extensions to
.Ar stdout
.It Fl n, --list-filenames
Prints all filenames in SOURCE
.It Fl c, --copy-regex Ar pattern
Copy all files from SOURCE matching
.Ar pattern
to a flat, specified output directory, set with
.Fl o
.It Fl d, --delete-regex Ar pattern
Delete all files matching a
.Ar pattern
.It Fl t, --trash-regex Ar pattern
Trash all files matching
.Ar pattern
to
.Pa ~/.local/Trash
.It Fl m, --move-regex Ar pattern
Moves all files matching
.Ar pattern
to a flat, specified output directory, set with
.Fl o
.It Fl o, --output
Sets the output directory; default is the current directory
.It Fl h, --help
Displays help
.It Fl -copy-stdin
Copies all files piped in through
.Ar stdin
into the output directory
.It Fl -delete-stdin
Deletes files piped in through
.Ar stdin
.It Fl -trash-stdin
Trashes files piped in through
.Ar stdin
to
.Pa ~/.local/Trash
.It Fl -move-stdin
Moves all files piped in through
.Ar stdin
into a flat output directory, set with
.Fl o
.It Fl -substitute-regex Ar matched-pattern,substituted-string
Copies files containing
.Ar matched-pattern
into an output directory set with
.Fl o
while preserving nesting and substituting all substrings
.Ar matched-pattern
with
.Ar substituted-string.
If the destructive flag is set, this operation will be done in-place.
.It Fl -clip-regex-end Ar pattern
Copies files ending in
.Ar pattern
into an output directory set with
.Fl o
while preserving nesting and truncating the ending
.Ar pattern.
If the destructive flag is set, this operation will be done in-place.
.It Fl -copy-extension Ar pattern
Copies all files ending in
.Ar pattern
to the a flat specified output directory set with
.Fl o.
The leading . is not automatically added to
.Ar pattern
.It Fl -delete-extension Ar pattern
Deletes files ending in
.Ar pattern.
The leading . is not automatically added to
.Ar pattern
.It Fl -trash-extension
Trashes files ending in
.Ar pattern
to
.Pa ~/.local/share/Trash.
The leading . is not automatically added to
.Ar pattern
.It Fl -move-extension
Moves all files ending in
.Ar pattern
to the a flat specified output directory set with
.Fl o.
The leading . is not automatically added to
.Ar pattern
.It Fl -destructive
Indicates that the source of the files does not need to be preserved. In
renaming or substitution operations, this will cause the files to be renamed
in place. In
.Fl f/--flatten,
this will automatically destroy the input directory.
.El
.Sh EXAMPLES
.Pp
An example is provided for each flag below, using the
.Pa ~/Pictures
directory as an example. Any instances of
.Fl o
and
.Fl -destructive
are optional and intended just as examples of their usage.
.Pp
Flatten a directory in-place (
.Fl f
):
.Dl ./manage.sh -o ~/Pictures --destructive -f ~/Pictures
List all extensions (
.Fl e
):
.Dl ./manage.sh -e ~/Pictures
List all filenames (
.Fl n
):
.Dl ./manage.sh -n ~/Pictures
Copy files matching a regex (
.Fl c
):
.Dl ./manage.sh -o ~/Pictures -c ".*\.jpg" ~/Pictures
Permanently delete files matching a regex (
.Fl d
):
.Dl ./manage.sh -d ".*\.jpg" ~/Pictures
Trash files matching a regex (
.Fl t
):
.Dl ./manage.sh -t ".*\.jpg" ~/Pictures
Move files matching a regex (
.Fl t
):
.Dl ./manage.sh -o ~/Pictures -m ".*\.jpg" ~/Pictures
Copy files piped from standard input:
.Dl find . -type f | ./manage.sh -o ~/Pictures --copy-stdin
Permanently delete files piped from standard input:
.Dl find . -type f | ./manage.sh --delete-stdin
Trash files piped from standard input:
.Dl find . -type f | ./manage.sh --trash-stdin
Move files piped from standard input:
.Dl find . -type f | ./manage.sh -o ~/Pictures --move-stdin
Rename files with regular expressions:
.Dl ./manage.sh -o ~/Pictures --substitute-regex "\.jpg,.png" ~/Downloads
Remove specific file endings:
.Dl ./manage.sh -o ~/Pictures --clip-regex-end "\.jpg" ~/Downloads
Copy files with specific file endings:
.Dl ./manage.sh -o ~/Pictures --copy-extension "\.jpg" ~/Downloads
Permanently delete files with specific file endings:
.Dl ./manage.sh --delete-extension "\.jpg" ~/Downloads
Trash files with specific file endings:
.Dl ./manage.sh --delete-extension "\.jpg" ~/Downloads
Move files with specific file endings:
.Dl ./manage.sh -o ~/Pictures --copy-extension "\.jpg" ~/Downloads
.Sh SEE ALSO
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr cp 1 ,
.Xr find 1 ,
.Xr mv 1 ,
.Xr sed 1 ,
.Xr xargs 1 ,


manage.sh
#!/bin/bash

OUTPUT_DIR="."
INPUT_DIR="${@: -1}"

IS_DESTRUCTIVE=false

VALID_ARGS=$(getopt -o "fenc:d:t:m:o:h"\
                    --long "copy-stdin,delete-stdin,trash-stdin,move-stdin,\
                    clip-regex-end:,delete-extension:,trash-extension:,\
                    copy-extension:,move-extension:,flatten-directory,\
                    list-extensions,list-filenames,copy-regex:,delete-regex:,\
                    trash-regex:,move-regex:,substitute-regex:,output:,help,\
                    destructive" -- "$@")

if [[ $? -ne 0 ]]; then
    exit 1;
fi

eval set -- "$VALID_ARGS"

while true; do
    case "$1" in
        -f | --flatten-directory)
            if [[ $IS_DESTRUCTIVE == true ]]; then
                find $INPUT_DIR -type f -print0 | xargs -0 -I{} mv {} $OUTPUT_DIR
                if [[ $INPUT_DIR != $OUTPUT_DIR ]]; then
                    rm -r $INPUT_DIR
                fi
            else
                find $INPUT_DIR -type f -print0 | xargs -0 -I{} cp {} $OUTPUT_DIR
            fi
            shift
            continue
            ;;
        -e | --list-extensions)
            # Passes nested directory listing through two regexes:
            # 1. Detect lines with filename extensions. Looks for pattern .*(\.[^\./]+)$, and takes only the paranthesized region
            # If the first regex fails (i.e. the file has no extension), a second regex is used
            # 2. Detect filenames (all files that make it to this step will have no extensions). Looks for pattern .*/([^\./]+)$
            find $INPUT_DIR -type f | sed -e "s/.*\(\.[^\.\/]\+\)$/\1/" | sed -e "s/.*\/\([^\/\.]\+\)$/\1/" | sort | uniq -c
            shift
            continue
            ;;
        -n | --list-filenames)
            find $INPUT_DIR | sed -e "s/^.*\///" | sort | uniq -c
            shift
            continue
            ;;
        -c | --copy-regex)
            find $INPUT_DIR -regex $2 -type f -print0 | xargs -0 -I{} cp {} $OUTPUT_DIR
            shift 2
            continue
            ;;
        -d | --delete-regex)
            find $INPUT_DIR -regex $2 -type f -print0 | xargs -0 -I{} rm {}
            shift 2
            continue
            ;;
        -t | --trash-regex)
            mkdir -p ~/.local/share/Trash
            find $INPUT_DIR -regex $2 -type f -print0 | xargs -0 -I{} mv {} ~/.local/share/Trash
            shift 2
            continue
            ;;
        -m | --move-regex)
            find $INPUT_DIR -regex $2 -type f -print0 | xargs -0 -I{} mv {} $OUTPUT_DIR
            shift 2
            continue
            ;;
        -o | --output)
            OUTPUT_DIR="$2"
            shift 2
            continue
            ;;
        -h | --help)
            echo "USAGE: manage.sh [OPTIONS]... SOURCE"
            echo "  or:  manage.sh -o DEST [OPTIONS]... SOURCE"
            echo "Perform OPTION(s) on SOURCE directory, and prefer output to DEST if applicable"
            echo "Multiple operations can be performed in a single invocation. Flags will be"
            echo "evaluated in order."
            echo "See the man page for more detail."
            shift
            continue
            ;;
        --substitute-regex)
            IFS=',' read -ra regexes <<< $2
            readarray -d $'\0' -t filenames < <(find $INPUT_DIR -regex ".*${regexes[0]}.*" -type f -print0)
            for filename in "${filenames[@]}"; do
                newfilename=$(basename $filename | sed -e "s/${regexes[0]}/${regexes[1]}/")
                if [[ $IS_DESTRUCTIVE == true ]]; then
                    targetdir="$(dirname $filename)"
                    mv $filename "$targetdir/$newfilename"
                else
                    targetdir="$(realpath $OUTPUT_DIR)/$(realpath $(dirname $filename))"
                    mkdir -p $targetdir
                    cp $filename "$targetdir/$newfilename"
                fi
            done
            shift 2
            continue
            ;;
        --copy-stdin)
            xargs -I{} cp {} $OUTPUT_DIR
            shift
            continue
            ;;
        --delete-stdin)
            xargs -I{} rm {}
            shift
            continue
            ;;
        --trash-stdin)
            mkdir -p ~/.local/share/Trash/
            xargs -I{} mv {} ~/.local/share/Trash/
            shift
            continue
            ;;
        --move-stdin)
            xargs -I{} mv {} $OUTPUT_DIR
            shift
            continue
            ;;
        --clip-regex-end)
            readarray -d $'\0' -t filenames < <(find $INPUT_DIR -regex ".*$2" -type f -print0)
            for filename in "${filenames[@]}"; do
                newfilename=$(basename $filename | sed -e "s/$2//")
                if [[ $IS_DESTRUCTIVE == true ]]; then
                    targetdir="$(dirname $filename)"
                    mv $filename "$targetdir/$newfilename"
                else
                    targetdir="$(realpath $OUTPUT_DIR)/$(realpath $(dirname $filename))"
                    mkdir -p $targetdir
                    cp $filename "$targetdir/$newfilename"
                fi
            done
            shift 2
            continue
            ;;
        --delete-extension)
            find $INPUT_DIR -iregex ".*$2" -type f -print0 | xargs -0 rm
            shift 2
            continue
            ;;
        --trash-extension)
            mkdir -p ~/.local/share/Trash/
            find $INPUT_DIR -iregex ".*$2" -type f -print0 | xargs -0 -I{} mv {} ~/.local/share/Trash/
            shift 2
            continue
            ;;
        --copy-extension)
            find $INPUT_DIR -iregex ".*$2" -type f -print0 | xargs -0 -I{} cp {} $OUTPUT_DIR
            shift 2
            continue
            ;;
        --move-extension)
            find $INPUT_DIR -iregex ".*$2" -type f -print0 | xargs -0 -I{} mv {} $OUTPUT_DIR
            shift 2
            ;;
        --destructive)
            IS_DESTRUCTIVE="true"
            shift 1
            ;;
        *)
            shift
            break
            ;;
    esac
done
